app/assets/javascripts/timeout.js

```diff
@@ -0,0 +1,149 @@
+const inactivity_mins = 28; // Minutes until first popup displayed
+const initial_warning_mins = 2; // Minutes until logout if no response after popup displayed
+const timeout_warning_mins = 58; // Minutes until popup displayed warning user of end of session
+
+(function (window) {
+  "use strict";
+
+  function isLoggedIn() {
+    // Checking whether user logged in or not based on whether logged_in_at exists
+    let timestamp;
+    if (typeof logged_in_at == 'undefined' || logged_in_at == 'None') {
+      timestamp =  'Test';
+    } else {
+      timestamp = new Date(logged_in_at);
+    }
+    return (timestamp instanceof Date) & !isNaN(timestamp);
+  }
+  function loggedInAt() {
+    if (typeof logged_in_at == 'undefined') {
+      return new Date(); // This needs to be created for the tests
+    } else {
+      return new Date(logged_in_at);
+    }
+  }
+
+  function getTimeLeftUntilExpiry() {
+    // calculates how long left in session, before final timeout, after each request
+    let expiry_time = new Date();
+    expiry_time.setMinutes(
+      loggedInAt().getMinutes() + timeout_warning_mins
+    );
+    let current_time = new Date();
```

Comment:
I'm not amazingly happy with doing date maths this way even though the documentation does say that values outside of 0-60 will update the date accordingly.  Perhaps consider using a proper date mathematic library like moment?

It would probably simplify quite a bit of the rest of the code too.